### PR TITLE
Make push rules use proper structures.

### PR DESCRIPTION
**Pull request by erikjohnston**
There are two major changes here:
1. replace using dicts for push rules with proper structs; and
2. stop copying the base rules all over the place.

I didn't split this into two separate PRs, as just doing the first point lead to a lot of effort that would later be removed by the second point. (LMK if I should do that, but the changes are pretty much a rewrite so an iterative approach may not be that helpful).

This improves load times for push rules:

| Version  | Time per user | Time for 1k users | 
| ------------- | ------------- | -- |
| Develop  | 138 µs  | 138ms |
| PR (with custom push rules)   | 2.11 µs  |  2.11ms |
| PR (without custom push rules) | 49.7 ns | 0.05 ms |

This therefore has a large impact on send times for rooms with large numbers of local users in the room.

Reviewable commit-by-commit (Note: you'll want to turn off whitespace for the second commit when comparing the change in base rules). There is also a comment commit at the end explaining wtf is going on. The changes to `base_rule.py` is pretty much a rewrite, so the diff isn't helpful.

### CODE REVIEW

File: synapse/storage/databases/main/push_rule.py [hunk 1]
```diff
@@ -859,8 +847,11 @@ async def copy_push_rules_from_room_to_room_for_user(
         user_push_rules = await self.get_push_rules_for_user(user_id)
 
         # Get rules relating to the old room and copy them to the new room
-        for rule in user_push_rules:
-            conditions = rule.get("conditions", [])
+        for rule, enabled in user_push_rules:
+            if not enabled:
+                continue
```
Comment (erikjohnston) [inline synapse/storage/databases/main/push_rule.py:852]: I think it was probably a bug that we were copying over custom disabled rules on room upgrades? Though I think in practice custom push rules get deleted rather than disabled.

File: synapse/push/baserules.py [hunk 2]
```diff
@@ -14,433 +14,544 @@
...
+    priority_class: int
+    conditions: Sequence[Mapping[str, str]]
+    actions: Sequence[Union[str, Mapping]]
+    default: bool = False
+    default_enabled: bool = True
 
-    Returns:
-        A new list with the rules set by the user combined with the defaults.
+
+@attr.s(auto_attribs=True, slots=True, frozen=True, weakref_slot=False)
```
Comment (clokep) [inline synapse/push/baserules.py:82]: Why the weakref here?

File: synapse/push/baserules.py [hunk 3]
```diff
@@ -14,433 +14,544 @@
...
+        ):
+            # Check if a base rule has been overriden by a custom rule. If so
+            # return that instead.
+            override_rule = self.overriden_base_rules.get(rule.rule_id)
+            if override_rule:
+                yield override_rule
+            else:
+                yield rule
+
+    def __len__(self) -> int:
```
Comment (clokep) [inline synapse/push/baserules.py:125]: Why does this not need the base rules?

File: synapse/push/baserules.py [hunk 4]
```diff
@@ -14,433 +14,544 @@
...
+        if rule.priority_class == 5:
+            collection = rules.override
+        elif rule.priority_class == 4:
+            collection = rules.content
+        elif rule.priority_class == 3:
+            collection = rules.room
+        elif rule.priority_class == 2:
+            collection = rules.sender
+        elif rule.priority_class == 1:
+            collection = rules.underride
```
Comment (clokep) [inline synapse/push/baserules.py:198]: Should the priority classes be an enum or constants?

Although I guess that would be a bigger change throughout this module, but probably worthwhile.

File: synapse/push/baserules.py [hunk 5]
```diff
@@ -14,172 +14,271 @@
...
-# `List[Dict[str, Sequence[Collection[str]]]]`.
-BASE_APPEND_CONTENT_RULES: List[Dict[str, Any]] = [
-    {
-        "rule_id": "global/content/.m.rule.contains_user_name",
-        "conditions": [
+def _is_experimental_rule_enabled(
+    rule_id: str, experimental_config: ExperimentalConfig
+) -> bool:
+    """Used by `_load_rules` to filter out experimental rules when they
+    have not been enabled.
```
Comment (clokep) [inline synapse/push/baserules.py:211]:
```suggestion
    """Used by `FilteredPushRules` to filter out experimental rules when they
    have not been enabled.
```

Although this makes me wonder if this should just be part of `FilteredPushRules`? No strong feelings either way.

File: synapse/push/clientformat.py [hunk 6]
```diff
@@ -18,28 +18,41 @@
...
-    for r in ruleslist:
-        template_name = _priority_class_to_template_name(r["priority_class"])
+    for r, enabled in ruleslist:
+        template_name = _priority_class_to_template_name(r.priority_class)
+
+        rulearray = rules["global"][template_name]
+
+        template_rule = _rule_to_template(r)
+        if not template_rule:
+            continue
```
Comment (clokep) [inline synapse/push/clientformat.py:44]: This change looks unrelated to the changes to attrs. Is this avoiding doing work that we don't care about if the template rule isn't found anyway?

File: synapse/push/clientformat.py [hunk 7]
```diff
@@ -18,28 +18,41 @@
...
+
+        rulearray.append(template_rule)
+
+        template_rule["enabled"] = enabled
+
+        if "conditions" not in template_rule:
+            continue
```
Comment (clokep) [inline synapse/push/clientformat.py:51]: We previously always had `conditions` and we don't now?

File: synapse/push/clientformat.py [hunk 8]
```diff
@@ -71,24 +74,24 @@ def _add_empty_priority_class_arrays(d: Dict[str, list]) -> Dict[str, list]:
     return d
 
 
-def _rule_to_template(rule: Dict[str, Any]) -> Optional[Dict[str, Any]]:
-    unscoped_rule_id = None
-    if "rule_id" in rule:
-        unscoped_rule_id = _rule_id_from_namespaced(rule["rule_id"])
```
Comment (clokep) [inline synapse/push/clientformat.py:77]: Did everything always previously have a rule ID anyway?

File: synapse/storage/databases/main/push_rule.py [hunk 9]
```diff
@@ -50,60 +62,31 @@
...
-    for i, rule in enumerate(rules):
-        rule_id = rule["rule_id"]
+        ruleslist.append(
+            PushRule(
+                rule_id=rawrule["rule_id"],
+                priority_class=rawrule["priority_class"],
+                conditions=db_to_json(rawrule["conditions"]),
+                actions=db_to_json(rawrule["actions"]),
+            )
+        )
```
Comment (clokep) [inline synapse/storage/databases/main/push_rule.py:83]: I think you could do this as a comprehension if you want now, theoretically that's slightly faster?

```suggestion
    ruleslist = [
        PushRule(
            rule_id=rawrule["rule_id"],
            priority_class=rawrule["priority_class"],
            conditions=db_to_json(rawrule["conditions"]),
            actions=db_to_json(rawrule["actions"]),
        ) for rawrule in rawrules
     ]
```

File: synapse/storage/databases/main/push_rule.py [hunk 10]
```diff
@@ -839,21 +812,27 @@ async def copy_push_rule_from_room_to_room(
...
 
+        new_conditions = []
+
         # Change room id in each condition
-        for condition in rule.get("conditions", []):
+        for condition in rule.conditions:
+            new_condition = condition
             if condition.get("key") == "room_id":
-                condition["pattern"] = new_room_id
+                new_condition = dict(condition)
```
Comment (clokep) [inline synapse/storage/databases/main/push_rule.py:824]: Is taking a copy of the condition here a bugfix?

File: synapse/storage/databases/main/push_rule.py [hunk 11]
```diff
@@ -827,21 +812,27 @@ async def copy_push_rule_from_room_to_room(
...
+                new_condition["pattern"] = new_room_id
+
+            new_conditions.append(condition)
```
Comment (clokep) [inline synapse/storage/databases/main/push_rule.py:827]: I think this is wrong:

```suggestion
            new_conditions.append(new_condition)
```

File: synapse/storage/databases/main/push_rule.py [hunk 12]
```diff
@@ -174,7 +145,7 @@ def get_max_push_rules_stream_id(self) -> int:
         raise NotImplementedError()
 
     @cached(max_entries=5000)
-    async def get_push_rules_for_user(self, user_id: str) -> List[JsonDict]:
+    async def get_push_rules_for_user(self, user_id: str) -> FilteredPushRules:
```
Comment (clokep) [inline synapse/storage/databases/main/push_rule.py:148]: Do we still need to do the sorting by priority class / priority here? (Isn't this handled by `PushRules` now?)

File: synapse/push/baserules.py [hunk 13]
```diff
@@ -14,172 +14,271 @@
...
-    # in the default positions in the list.
-    rawrules = [r for r in rawrules if r["priority_class"] >= 0]
 
-    # shove the server default rules for each kind onto the end of each
-    current_prio_class = list(PRIORITY_CLASS_INVERSE_MAP)[-1]
+@attr.s(auto_attribs=True, slots=True, frozen=True, weakref_slot=False)
+class FilteredPushRules:
+    """A wrapper around `PushRules` that filters out disabled experimental push
+    rules, and includes the "enabled" state for each rule when iterated over.
+    """
```
Comment (clokep) [inline synapse/push/baserules.py:140]: I don't think `PushRules` is ever directly used, which makes me wonder if there's any use in it being a separate object?

File: synapse/push/clientformat.py (hunk 8 above)
Comment (clokep) [reply synapse/push/clientformat.py:77]: I.e. was this if check useless?

File: synapse/push/clientformat.py (hunk 7 above)
Comment (clokep) [reply synapse/push/clientformat.py:51]: Hmm or is this actually a bugfix and assuming we always had `conditions` wasn't valid?

File: synapse/push/baserules.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/push/baserules.py:82]: Attrs automatically adds a slot for weakrefs, which we don't use so I idly turned it off :shrug:

File: synapse/push/baserules.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/push/baserules.py:125]: It's mostly used by the cache to know how "large" in memory the `PushRules` are, and the base rules don't count towards that. Will comment.

File: synapse/push/baserules.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/push/baserules.py:198]: Yeah, probably. But I don't want to make this PR bigger than it is.

File: synapse/push/baserules.py (hunk 5 above)
Comment (erikjohnston) [reply synapse/push/baserules.py:211]: Yeah, good point. Though I think I'm going to keep it separate since it feels like it might get used elsewhere potentially.

File: synapse/push/clientformat.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/push/clientformat.py:44]: I found it easier to get the `template_rule` early and then update that, rather than what was here before which deep copied the push rules and mutated them in place.

File: synapse/push/clientformat.py (hunk 8 above)
Comment (erikjohnston) [reply synapse/push/clientformat.py:77]: Yeah, the DB has a not null check for `rule_id` for example.

File: synapse/storage/databases/main/push_rule.py (hunk 10 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/push_rule.py:824]: Oh, err, maybe? I thought we copied the push rule by this point but I can't find where we do that now

File: synapse/storage/databases/main/push_rule.py (hunk 11 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/push_rule.py:827]: Whoops

File: synapse/storage/databases/main/push_rule.py (hunk 12 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/push_rule.py:148]: The `PushRules` class will correctly sort by priority class, but doesn't have a concept of "priority" within a class, so expects to get the push rules in the correct order

File: synapse/push/clientformat.py (hunk 7 above)
Comment (erikjohnston) [reply synapse/push/clientformat.py:51]: Not all rules have conditions when they've been formatted, e.g. room rules don't have conditions as they can be inferred from the kind and rule ID. Will add a comment.

(Note at the `continue` we've already added the rule to the result list, we're just skipping the formatting of the conditions section).

File: synapse/push/baserules.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/push/baserules.py:125]: Hmm, actually that is pretty confusing to have a `__len__` which gives a different result than `__iter__`. Not quite sure how to get out of that? I guess we could finally add support to the cache descriptors to let you specify a function?

File: synapse/push/baserules.py (hunk 13 above)
Comment (erikjohnston) [reply synapse/push/baserules.py:143]: Hmm, maybe. I'm a bit torn about it TBH, because I think its weird that we're caching `FilteredPushRules` rather than `PushRules`, given we also cache the `get_push_rules_enabled_for_user` stuff.

Comment (clokep) [reply synapse/push/baserules.py:143]: I don't feel strongly about it, just seemed like an extra layer of indirection and if we're trying to reduce the number of objects created it would be useful. 🤷 I think keeping it separate is probably fine.

File: synapse/push/clientformat.py (hunk 6 above)
Comment (clokep) [reply synapse/push/clientformat.py:44]: That makes sense -- I guess my thought is this isn't inherent to the change to structures. 👍

File: synapse/push/baserules.py (hunk 3 above)
Comment (clokep) [reply synapse/push/baserules.py:125]: I think documenting it for now is fine. 👍

**Review by clokep — APPROVED** [review]

**Comment by erikjohnston** [comment]
Woo, thanks!

File: synapse/push/clientformat.py (hunk 6 above)
Comment (DMRobertson) [reply synapse/push/clientformat.py:44]: I think this caused #14348 because the cleanup logic ("# Remove internal stuff" below)is not longer applied before `_rule_to_template` is called.
